src/pw/fpga/opencl_utils.c

```diff
@@ -0,0 +1,370 @@
+ * \param   number of devices found
+ * \param   size of FFT3d
+ * \retval  created program
+ *****************************************************************************/
+cl_program getProgramWithBinary(cl_context context, const cl_device_id *devices, unsigned num_device, int N[3], char *data_path){
+  char bin_path[500];
+  char *binary;
+  char *binaries[num_device];
+
+  size_t bin_size;
+  cl_int bin_status, status;
+
+  strcpy(bin_path, data_path);
+
+  // Get the full path of the binary
+  if( !getBinaryPath(bin_path, N) ){
+    printf("No paths to the binary found\n");
+    return NULL;
+  }
+
+  // Load binary to character array
+  binary = loadBinary(bin_path, &bin_size);
+  if(binary == NULL){
+    printf("Could not load binary\n");
+    free(binary);
+    return NULL;
+  }
+
+  binaries[0] = binary;
+
+  // Create the program.
+  cl_program program = clCreateProgramWithBinary(context, 1, devices, &bin_size, (const unsigned char **) binaries, &bin_status, &status);
+  if (status != CL_SUCCESS){
+    printf("Query to create program with binary failed\n");
+    free(binary);
+    return NULL;
+  }
+  free(binary);
+  return program;
+}
+
+static bool fileTestExists(char* filename){
+  if( access( filename, R_OK ) != -1 ) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+static bool getBinaryPath(char *path, int N[3]){
+
+#ifdef __PW_FPGA_SP
+  char *subpath = "/../fpgabitstream/fft3d/synthesis_sp/";
+#else
+  char *subpath = "/../fpgabitstream/fft3d/synthesis_dp/";
+#endif
+  
+  switch(N[0]){
+    case 16 : 
+      strcat(path, subpath);
+      strcat(path, "syn16/");
+      break;
+
+    case 32 : 
+      strcat(path, subpath);
+      strcat(path, "syn32/");
+      break;
+
+    case 64 : 
+      strcat(path, subpath);
+      strcat(path, "syn64/");
+      break;
+
+    default:
+      printf("Path not found for the size (%d,%d,%d)", N[0], N[1], N[2]);
+      exit(0);
+      break;
+  }
+  // strcat(full_path, specific_path); // transfer specific path to filename 
+  strcat(path, "fft3d");      // Append filename
+  strcat(path, ".aocx");      // Append extension
+  strcat(path, "\0");         // Append extension
+
+  if (!fileTestExists(path)){
+    printf("File not found \n");
+    return false;
+  }
+
+  return true;
+}
+
+static char* loadBinary(char *binary_path, size_t *bin_size){
+
+  FILE *fp;
+
+  // Open file and check if it exists
+  fp = fopen(binary_path, "rb");
+  if(fp == 0){
+    return NULL;
+  }
+
+  // Find the size of the file
+  fseek(fp, 0L, SEEK_END);
+  *bin_size = ftell(fp);
+
+  char *binary = (char *)malloc(*bin_size);
+  rewind(fp);
+
+  if(fread( (void *)binary, *bin_size, 1, fp) == 0) {
+    free(binary);
+    fclose(fp);
+    return NULL;
+  }
+
+  fclose(fp);
+  return binary;
+}
+
+// Minimum alignment requirement to use DMA
+const unsigned OPENCL_ALIGNMENT = 64;
+void* alignedMalloc(size_t size){
+
+  void *memptr = NULL;
+  int ret = posix_memalign(&memptr, OPENCL_ALIGNMENT, size);
```

Comment:
This temporary can be removed, instead you might want to use something like https://stackoverflow.com/a/26855379 instead (respectively a simple wrapper around it).

It might also be good to document the default behavior of `posix_memalign` to leave `memptr` untouched or even explicitly set to `NULL`, leaving it to the caller to make sure the pointer returned is not `NULL`